src/bin/pg_autoctl/file_utils.c

```diff
@@ -119,6 +119,35 @@ ensure_empty_dir(const char *dirname, int mode)
 }
 
 
+/*
+ * fopen_with_umask is a version of fopen that gives more control. The main
+ * advantage of it is that it allows specifying a umask of the file. This makes
+ * sure files are not accidentally created with umask 777 if the user has it
+ * configured in a weird way.
+ *
+ * This function returns NULL when opening the file fails. So this should be
+ * handled. It will log an error in this case though, so that's not necessary
+ * at the callsite.
+ */
+FILE *
+fopen_with_umask(const char *filePath, const char* modes, int flags, mode_t umask)
+{
+	int fileDescriptor = open(filePath, O_WRONLY | O_CREAT, umask);
+	FILE *fileStream = NULL;
+	if (fileDescriptor != -1)
+	{
+		fileStream = fdopen(fileDescriptor, modes);
+		if (fileStream == NULL) {
+			close(fileDescriptor);
+		}
+	}
+	if (fileStream == NULL) {
+		log_error("Failed to open file \"%s\": %s", filePath, strerror(errno));
+	}
```

Comment:
Style issue: brackets on their own lines. Also, more to the point, do we need to check if fileStream is NULL in two different places? We could log_error and return from the first place, I guess?